db/fetcher/indexer_iterators.go

```diff
@@ -157,38 +160,33 @@ func (i *eqSingleIndexIterator) Close() error {
 }
 
 type inIndexIterator struct {
-	filterValueIndexIterator
-	filterValues [][]byte
-	nextValIndex int
-	ctx          context.Context
-	store        datastore.DSReaderWriter
-	hasIterator  bool
-}
-
-func newInIndexIterator(
-	indexIter filterValueIndexIterator,
-	filterValues [][]byte,
-) *inIndexIterator {
-	return &inIndexIterator{
-		filterValueIndexIterator: indexIter,
-		filterValues:             filterValues,
-	}
+	indexIterator
+	keyFieldValues [][]byte
+	nextValIndex   int
+	ctx            context.Context
+	store          datastore.DSReaderWriter
+	hasIterator    bool
 }
 
 func (i *inIndexIterator) nextIterator() (bool, error) {
 	if i.nextValIndex > 0 {
-		err := i.filterValueIndexIterator.Close()
+		err := i.indexIterator.Close()
 		if err != nil {
 			return false, err
 		}
 	}
 
-	if i.nextValIndex >= len(i.filterValues) {
+	if i.nextValIndex >= len(i.keyFieldValues) {
 		return false, nil
 	}
 
-	i.SetFilterValue(i.filterValues[i.nextValIndex])
-	err := i.filterValueIndexIterator.Init(i.ctx, i.store)
+	switch fieldIter := i.indexIterator.(type) {
+	case *eqPrefixIndexIterator:
+		fieldIter.keyFieldValue = i.keyFieldValues[i.nextValIndex]
+	case *eqSingleIndexIterator:
+		fieldIter.keyFieldValues[0] = i.keyFieldValues[i.nextValIndex]
```

Comment:
question: I don't see a constructor for `eqSingleIndexIterator`, where is it guaranteed that the slice will have a `0` index?

Reply:
it is created in `createIndexIterator` and it's guaranteed because it's the field value we are seeking and the value is being  set upon creation.

Reply:
Ah okay, I would prefer a constructor for stuff like that, but `createIndexIterator` is a factory-like func and  the concrete a private type, so I am happy for this to be shrugged off as my personal preference here :grin: